graphql_compiler/schema_transformation/merge_schemas.py

```diff
@@ -0,0 +1,173 @@
+            ast_types.ObjectTypeDefinition(
+                name=ast_types.Name(value=query_type),
+                fields=[],
+                interfaces=[],
+                directives=[],
+            ),
+        ]
+    )
+    return blank_ast
+
+
+def merge_schemas(schemas_dict):
+    """Check that input schemas do not contain conflicting definitions, then merge.
+
+    Merged schema will contain all type, interface, enum, scalar, and directive definitions
+    from input schemas. The fields of its query type will be the union of the fields of the
+    query types of each input schema.
+
+    Args:
+        schemas_dict: OrderedDict[str, Document], where keys are names/identifiers of schemas,
+                      and values are ASTs describing schemas. The ASTs will not be modified
+                      by this funcion
+
+    Returns:
+        MergedSchemaDescriptor, a namedtuple that contains the AST of the merged schema,
+        and the map from names of types/query type fields to the id of the schema that they
+        came from
+
+    Raises:
+        - SchemaStructureError if the schema does not have the expected form; in particular, if
+          the AST does not represent a valid schema, if any query type field does not have the
+          same name as the type that it queries, if the schema contains type extensions or
+          input object definitions, or if the schema contains mutations or subscriptions
+        - SchemaNameConflictError if there are conflicts between the names of
+          types/interfaces/enums/scalars, or conflicts between the definition of directives
+          with the same name
+    """
+    if len(schemas_dict) == 0:
+        raise ValueError(u'Expected a nonzero number of schemas to merge.')
+
+    query_type = 'RootSchemaQuery'
+    merged_schema_ast = _basic_schema_ast(query_type)  # Document
+
+    name_to_schema_id = {}  # Dict[str, str], name of type/interface/enum/union to schema id
+    scalars = {'String', 'Int', 'Float', 'Boolean', 'ID'}  # Set[str], user defined + builtins
+    directives = {}  # Dict[str, DirectiveDefinition]
+
+    for cur_schema_id, cur_ast in six.iteritems(schemas_dict):
+        # Prevent aliasing between output and input
+        cur_ast = deepcopy(cur_ast)
+
+        # Check input schema satisfies various structural requirements
+        check_ast_schema_is_valid(cur_ast)
+
+        cur_schema = build_ast_schema(cur_ast)
+        cur_query_type = get_query_type_name(cur_schema)
+
+        # Merge cur_ast into merged_schema_ast
+        # Concatenate new scalars, new directives, and type definitions other than the query
+        # type to definitions list
+        # Raise errors for conflicting scalars, directives, or types
+        new_definitions = cur_ast.definitions  # List[Node]
+        new_query_type_fields = None  # List[FieldDefinition]
+        for new_definition in new_definitions:
+            if isinstance(new_definition, ast_types.SchemaDefinition):
```

Comment:
Is there anything we'd like to validate on this definition before skipping to the next one?

Reply:
The SchemaDefinition looks something like this:
`SchemaDefinition(operation_types=[OperationTypeDefinition(operation='query', type=NamedType(name=Name(value='SchemaQuery')))], directives=[])`
`check_ast_schema_is_valid` already confirms that the SchemaDefinition only contains a query operation. It doesn't feel like there's a lot else to check. I supposed I could maybe check that the list of directives is empty (perhaps I should put that check into `check_ast_schema_is_valid`)?

Reply:
It's probably fine as it is, I just wanted to confirm with you since you are a lot more familiar with the AST structure than me.